In [73]:
import pandas as pd
import numpy as np
import time

In [74]:
df = pd.read_csv("D:\\DAUSem1\\ML\\Dataset\\garments_worker_productivity.csv")
df.head()

,date,quarter,department,day,team,targeted_productivity,smv,wip,over_time,incentive,idle_time,idle_men,no_of_style_change,no_of_workers,actual_productivity
0,1/1/2015,Quarter1,sweing,Thursday,8,0.80,26.16,1108.0,7080,98,0.0,0,0,59.0,0.940725
1,1/1/2015,Quarter1,finishing,Thursday,1,0.75,3.94,NaN,960,0,0.0,0,0,8.0,0.886500
2,1/1/2015,Quarter1,sweing,Thursday,11,0.80,11.41,968.0,3660,50,0.0,0,0,30.5,0.800570
3,1/1/2015,Quarter1,sweing,Thursday,12,0.80,11.41,968.0,3660,50,0.0,0,0,30.5,0.800570
4,1/1/2015,Quarter1,sweing,Thursday,6,0.80,25.90,1170.0,1920,50,0.0,0,0,56.0,0.800382


In [75]:
df.shape

(1197, 15)

In [76]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1197 entries, 0 to 1196
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date                   1197 non-null   str    
 1   quarter                1197 non-null   str    
 2   department             1197 non-null   str    
 3   day                    1197 non-null   str    
 4   team                   1197 non-null   int64  
 5   targeted_productivity  1197 non-null   float64
 6   smv                    1197 non-null   float64
 7   wip                    691 non-null    float64
 8   over_time              1197 non-null   int64  
 9   incentive              1197 non-null   int64  
 10  idle_time              1197 non-null   float64
 11  idle_men               1197 non-null   int64  
 12  no_of_style_change     1197 non-null   int64  
 13  no_of_workers          1197 non-null   float64
 14  actual_productivity    1197 non-null   float64
dtypes: float64(6), 

In [77]:
df.isnull().sum()

date                       0
quarter                    0
department                 0
day                        0
team                       0
targeted_productivity      0
smv                        0
wip                      506
over_time                  0
incentive                  0
idle_time                  0
idle_men                   0
no_of_style_change         0
no_of_workers              0
actual_productivity        0
dtype: int64

In [78]:
df.describe()

,team,targeted_productivity,smv,wip,over_time,incentive,idle_time,idle_men,no_of_style_change,no_of_workers,actual_productivity
count,1197.000000,1197.000000,1197.000000,691.000000,1197.000000,1197.000000,1197.000000,1197.000000,1197.000000,1197.000000,1197.000000
mean,6.426901,0.729632,15.062172,1190.465991,4567.460317,38.210526,0.730159,0.369256,0.150376,34.609858,0.735091
std,3.463963,0.097891,10.943219,1837.455001,3348.823563,160.182643,12.709757,3.268987,0.427848,22.197687,0.174488
min,1.000000,0.070000,2.900000,7.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000,0.233705
25%,3.000000,0.700000,3.940000,774.500000,1440.000000,0.000000,0.000000,0.000000,0.000000,9.000000,0.650307
50%,6.000000,0.750000,15.260000,1039.000000,3960.000000,0.000000,0.000000,0.000000,0.000000,34.000000,0.773333
75%,9.000000,0.800000,24.260000,1252.500000,6960.000000,50.000000,0.000000,0.000000,0.000000,57.000000,0.850253
max,12.000000,0.800000,54.560000,23122.000000,25920.000000,3600.000000,300.000000,45.000000,2.000000,89.000000,1.120437


In [208]:
df.describe(include="str")

,quarter,department,day
count,1197,1197,1197
unique,5,3,6
top,Quarter1,sweing,Wednesday
freq,360,691,208


In [80]:
y_reg = df["actual_productivity"]
df["MeetsTarget"] = (
    df["actual_productivity"] >= df["targeted_productivity"]).astype(int)

df["MeetsTarget"].value_counts()

MeetsTarget
1    875
0    322
Name: count, dtype: int64

In [81]:
df[["targeted_productivity", "actual_productivity", "MeetsTarget"]].head(10)

,targeted_productivity,actual_productivity,MeetsTarget
0,0.80,0.940725,1
1,0.75,0.886500,1
2,0.80,0.800570,1
3,0.80,0.800570,1
4,0.80,0.800382,1
5,0.80,0.800125,1
6,0.75,0.755167,1
7,0.75,0.753683,1
8,0.75,0.753098,1
9,0.75,0.750428,1


In [82]:
X_reg = df.drop(columns=["actual_productivity", "MeetsTarget"])
y_reg = df["actual_productivity"]

X_cls = df.drop(columns=["actual_productivity", "MeetsTarget"])
y_cls = df["MeetsTarget"]

In [83]:
df.shape
df.info()
df.isnull().sum()
df.describe()
df["MeetsTarget"].value_counts()

<class 'pandas.DataFrame'>
RangeIndex: 1197 entries, 0 to 1196
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date                   1197 non-null   str    
 1   quarter                1197 non-null   str    
 2   department             1197 non-null   str    
 3   day                    1197 non-null   str    
 4   team                   1197 non-null   int64  
 5   targeted_productivity  1197 non-null   float64
 6   smv                    1197 non-null   float64
 7   wip                    691 non-null    float64
 8   over_time              1197 non-null   int64  
 9   incentive              1197 non-null   int64  
 10  idle_time              1197 non-null   float64
 11  idle_men               1197 non-null   int64  
 12  no_of_style_change     1197 non-null   int64  
 13  no_of_workers          1197 non-null   float64
 14  actual_productivity    1197 non-null   float64
 15  MeetsTarget    

MeetsTarget
1    875
0    322
Name: count, dtype: int64

In [84]:
df["date"] = pd.to_datetime(df["date"])

df["year"] = df["date"].dt.year
df["month"] = df["date"].dt.month
df["day_of_month"] = df["date"].dt.day

In [85]:
df = df.drop(columns=["date"])

In [86]:
X_reg = df.drop(columns=["actual_productivity", "MeetsTarget"])

y_reg = df["actual_productivity"]

In [87]:
X_cls = df.drop(columns=["actual_productivity", "MeetsTarget"])

y_cls = df["MeetsTarget"]

In [88]:
np.random.seed(42)

indices = np.arange(len(df))
np.random.shuffle(indices)

split_point = int(0.8 * len(indices))

train_idx = indices[:split_point]
test_idx = indices[split_point:]

In [89]:
len(train_idx), len(test_idx)

(957, 240)

In [90]:
print("Train class distribution:")
print(y_cls.iloc[train_idx].value_counts(normalize=True))

print("\nTest class distribution:")
print(y_cls.iloc[test_idx].value_counts(normalize=True))

Train class distribution:
MeetsTarget
1    0.735632
0    0.264368
Name: proportion, dtype: float64

Test class distribution:
MeetsTarget
1    0.7125
0    0.2875
Name: proportion, dtype: float64


In [91]:
df.shape
X_reg.shape
X_cls.shape

print("Train:", len(train_idx))
print("Test:", len(test_idx))

print("\nTrain class distribution:")
print(y_cls.iloc[train_idx].value_counts(normalize=True))

print("\nTest class distribution:")
print(y_cls.iloc[test_idx].value_counts(normalize=True))

Train: 957
Test: 240

Train class distribution:
MeetsTarget
1    0.735632
0    0.264368
Name: proportion, dtype: float64

Test class distribution:
MeetsTarget
1    0.7125
0    0.2875
Name: proportion, dtype: float64


## Part A: Scikit-learn preprocessing

In [92]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

categorical_cols = [
    "quarter",
    "department",
    "day"
]

numeric_cols = [
    "team",
    "targeted_productivity",
    "smv",
    "wip",
    "over_time",
    "incentive",
    "idle_time",
    "idle_men",
    "no_of_style_change",
    "no_of_workers",
    "year",
    "month",
    "day_of_month"
]

In [93]:
numeric_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

categorical_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

In [94]:
preprocessor = ColumnTransformer([
    ("num", numeric_pipeline, numeric_cols),
    ("cat", categorical_pipeline, categorical_cols)
])

In [95]:
X_reg_train = X_reg.iloc[train_idx]
X_reg_test = X_reg.iloc[test_idx]

y_reg_train = y_reg.iloc[train_idx]
y_reg_test = y_reg.iloc[test_idx]

In [96]:
X_cls_train = X_cls.iloc[train_idx]
X_cls_test = X_cls.iloc[test_idx]

y_cls_train = y_cls.iloc[train_idx]
y_cls_test = y_cls.iloc[test_idx]

In [97]:
print(X_reg_train.shape)
print(X_reg_test.shape)

print(X_cls_train.shape)
print(X_cls_test.shape)

(957, 16)
(240, 16)
(957, 16)
(240, 16)


In [98]:
X_reg_train_processed = preprocessor.fit_transform(X_reg_train)
X_reg_test_processed = preprocessor.transform(X_reg_test)

In [99]:
X_cls_train_processed = preprocessor.transform(X_cls_train)
X_cls_test_processed = preprocessor.transform(X_cls_test)

In [100]:
print("Original features:", X_reg_train.shape[1])
print("Processed features:", X_reg_train_processed.shape[1])

Original features: 16
Processed features: 27


In [101]:
from sklearn.linear_model import LinearRegression

reg_model = LinearRegression()
start_time = time.perf_counter()

reg_model.fit(
    X_reg_train_processed,
    y_reg_train
)

reg_train_time = time.perf_counter() - start_time

In [102]:
start_time = time.perf_counter()

y_reg_pred = reg_model.predict(
    X_reg_test_processed
)

reg_prediction_time = time.perf_counter() - start_time

In [103]:
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)


reg_mae = mean_absolute_error( y_reg_test,y_reg_pred)
reg_rmse = np.sqrt(mean_squared_error(y_reg_test,y_reg_pred))
reg_r2 = r2_score(y_reg_test,y_reg_pred)

In [104]:
print("Scikit-learn Linear Regression")
print("--------------------------------")
print("MAE :", reg_mae)
print("RMSE:", reg_rmse)
print("R2  :", reg_r2)

print("\nTraining time  :", reg_train_time)
print("Prediction time:", reg_prediction_time)

Scikit-learn Linear Regression
--------------------------------
MAE : 0.10513722880204522
RMSE: 0.14406347309170997
R2  : 0.32655105050411226

Training time  : 0.0029474999755620956
Prediction time: 0.00046499993186444044


In [105]:
reg_results = pd.DataFrame({
    "Model": ["Scikit-learn Linear Regression"],
    "MAE": [reg_mae],
    "RMSE": [reg_rmse],
    "R2": [reg_r2],
    "Training Time": [reg_train_time],
    "Prediction Time": [reg_prediction_time]
})

reg_results

,Model,MAE,RMSE,R2,Training Time,Prediction Time
0,Scikit-learn Linear Regression,0.105137,0.144063,0.326551,0.002947,0.000465


In [106]:
classification_preprocessor = ColumnTransformer([
    ("num", numeric_pipeline, numeric_cols),
    ("cat", categorical_pipeline, categorical_cols)
])

In [107]:
X_cls_train_processed = classification_preprocessor.fit_transform(
    X_cls_train
)

X_cls_test_processed = classification_preprocessor.transform(
    X_cls_test
)

In [108]:
print("Original:", X_cls_train.shape)
print("Processed:", X_cls_train_processed.shape)
print("Test:", X_cls_test_processed.shape)

Original: (957, 16)
Processed: (957, 27)
Test: (240, 27)


In [109]:
from sklearn.linear_model import LogisticRegression

cls_model = LogisticRegression(max_iter=1000)

start_time = time.perf_counter()

cls_model.fit(
    X_cls_train_processed,
    y_cls_train
)

cls_train_time = time.perf_counter() - start_time

start_time = time.perf_counter()

y_cls_pred = cls_model.predict(
    X_cls_test_processed
)

cls_prediction_time = time.perf_counter() - start_time

In [110]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

cls_accuracy = accuracy_score(
    y_cls_test,
    y_cls_pred
)

cls_precision = precision_score(
    y_cls_test,
    y_cls_pred
)

cls_recall = recall_score(
    y_cls_test,
    y_cls_pred
)

cls_f1 = f1_score(
    y_cls_test,
    y_cls_pred
)

In [111]:
print("Scikit-learn Logistic Regression")
print("----------------------------------")
print("Accuracy :", cls_accuracy)
print("Precision:", cls_precision)
print("Recall   :", cls_recall)
print("F1-score :", cls_f1)

print("\nTraining time  :", cls_train_time)
print("Prediction time:", cls_prediction_time)

Scikit-learn Logistic Regression
----------------------------------
Accuracy : 0.7625
Precision: 0.7766990291262136
Recall   : 0.935672514619883
F1-score : 0.8488063660477454

Training time  : 0.010901800007559359
Prediction time: 0.000308999908156693


In [112]:
cls_results = pd.DataFrame({
    "Model": ["Scikit-learn Logistic Regression"],
    "Accuracy": [cls_accuracy],
    "Precision": [cls_precision],
    "Recall": [cls_recall],
    "F1": [cls_f1],
    "Training Time": [cls_train_time],
    "Prediction Time": [cls_prediction_time]
})

cls_results

,Model,Accuracy,Precision,Recall,F1,Training Time,Prediction Time
0,Scikit-learn Logistic Regression,0.7625,0.776699,0.935673,0.848806,0.010902,0.000309


## Part B - From-Scratch Implementation

### Copying train and test features

In [113]:
X_reg_train_manual = X_reg_train.copy()
X_reg_test_manual = X_reg_test.copy()

X_cls_train_manual = X_cls_train.copy()
X_cls_test_manual = X_cls_test.copy()

In [114]:
numeric_medians = X_reg_train_manual[numeric_cols].median()

numeric_medians

team                        6.00
targeted_productivity       0.75
smv                        15.26
wip                      1047.00
over_time                4020.00
incentive                   0.00
idle_time                   0.00
idle_men                    0.00
no_of_style_change          0.00
no_of_workers              34.00
year                     2015.00
month                       2.00
day_of_month               12.00
dtype: float64

### FOR LINEAR REGRESSION

In [115]:
X_reg_train_manual[numeric_cols] = (
    X_reg_train_manual[numeric_cols]
    .fillna(numeric_medians)
)

X_reg_test_manual[numeric_cols] = (
    X_reg_test_manual[numeric_cols]
    .fillna(numeric_medians)
)

In [116]:
print("Missing values in train:")
print(X_reg_train_manual.isnull().sum())

print("\nMissing values in test:")
print(X_reg_test_manual.isnull().sum())

Missing values in train:
quarter                  0
department               0
day                      0
team                     0
targeted_productivity    0
smv                      0
wip                      0
over_time                0
incentive                0
idle_time                0
idle_men                 0
no_of_style_change       0
no_of_workers            0
year                     0
month                    0
day_of_month             0
dtype: int64

Missing values in test:
quarter                  0
department               0
day                      0
team                     0
targeted_productivity    0
smv                      0
wip                      0
over_time                0
incentive                0
idle_time                0
idle_men                 0
no_of_style_change       0
no_of_workers            0
year                     0
month                    0
day_of_month             0
dtype: int64


In [117]:
categorical_cols = ["quarter","department","day"]
category_values = {}

for col in categorical_cols:
    category_values[col] = X_reg_train_manual[col].unique()

category_values

{'quarter': <ArrowStringArray>
 ['Quarter4', 'Quarter3', 'Quarter1', 'Quarter2', 'Quarter5']
 Length: 5, dtype: str,
 'department': <ArrowStringArray>
 ['sweing', 'finishing ', 'finishing']
 Length: 3, dtype: str,
 'day': <ArrowStringArray>
 ['Tuesday', 'Monday', 'Sunday', 'Thursday', 'Saturday', 'Wednesday']
 Length: 6, dtype: str}

In [118]:
for col in categorical_cols:
    X_reg_train_manual[col] = X_reg_train_manual[col].str.strip()
    X_reg_test_manual[col] = X_reg_test_manual[col].str.strip()

In [119]:
for col in categorical_cols:
    print(col, X_reg_train_manual[col].unique())

quarter <ArrowStringArray>
['Quarter4', 'Quarter3', 'Quarter1', 'Quarter2', 'Quarter5']
Length: 5, dtype: str
department <ArrowStringArray>
['sweing', 'finishing']
Length: 2, dtype: str
day <ArrowStringArray>
['Tuesday', 'Monday', 'Sunday', 'Thursday', 'Saturday', 'Wednesday']
Length: 6, dtype: str


In [120]:
X_reg_train_manual = pd.get_dummies(
    X_reg_train_manual,
    columns=categorical_cols,
    dtype=float
)

X_reg_test_manual = pd.get_dummies(
    X_reg_test_manual,
    columns=categorical_cols,
    dtype=float
)

In [121]:
print("Train shape:", X_reg_train_manual.shape)
print("Test shape :", X_reg_test_manual.shape)

Train shape: (957, 26)
Test shape : (240, 26)


In [122]:
X_reg_test_manual = X_reg_test_manual.reindex(
    columns=X_reg_train_manual.columns,
    fill_value=0
)

In [123]:
print("Train shape:", X_reg_train_manual.shape)
print("Test shape :", X_reg_test_manual.shape)

print(
    "Same columns:",
    list(X_reg_train_manual.columns) == list(X_reg_test_manual.columns)
)

Train shape: (957, 26)
Test shape : (240, 26)
Same columns: True


### Calculating Mean

In [128]:
train_mean = X_reg_train_manual.mean()
train_std = X_reg_train_manual.std()

print(train_mean)

team                        6.387670
targeted_productivity       0.730376
smv                        15.128119
wip                      1094.859979
over_time                4602.183908
incentive                  36.688610
idle_time                   0.898119
idle_men                    0.367816
no_of_style_change          0.141066
no_of_workers              34.763323
year                     2015.000000
month                       1.705329
day_of_month               13.663532
quarter_Quarter1            0.292581
quarter_Quarter2            0.282132
quarter_Quarter3            0.180773
quarter_Quarter4            0.204807
quarter_Quarter5            0.039707
department_finishing        0.422153
department_sweing           0.577847
day_Monday                  0.169279
day_Saturday                0.149425
day_Sunday                  0.174504
day_Thursday                0.156740
day_Tuesday                 0.168234
day_Wednesday               0.181818
dtype: float64


### Calculating Standard Deviation

In [130]:
print(train_std)

team                        3.476938
targeted_productivity       0.097719
smv                        10.964951
wip                      1062.504942
over_time                3334.649694
incentive                 146.206663
idle_time                  14.208854
idle_men                    3.284700
no_of_style_change          0.416646
no_of_workers              22.185167
year                        0.000000
month                       0.739574
day_of_month                8.668805
quarter_Quarter1            0.455186
quarter_Quarter2            0.450272
quarter_Quarter3            0.385031
quarter_Quarter4            0.403771
quarter_Quarter5            0.195373
department_finishing        0.494161
department_sweing           0.494161
day_Monday                  0.375194
day_Saturday                0.356694
day_Sunday                  0.379740
day_Thursday                0.363745
day_Tuesday                 0.374270
day_Wednesday               0.385896
dtype: float64


In [132]:
X_reg_train_manual = (X_reg_train_manual - train_mean) / train_std

X_reg_test_manual = (X_reg_test_manual - train_mean) / train_std

In [136]:
print("Train mean after scaling:")
print(X_reg_train_manual.mean().round(3))

print("\nTrain standard deviation after scaling:")
print(X_reg_train_manual.std().round(3))

Train mean after scaling:
team                    -1.837
targeted_productivity   -7.474
smv                     -1.380
wip                     -1.030
over_time               -1.380
incentive               -0.251
idle_time               -0.063
idle_men                -0.112
no_of_style_change      -0.339
no_of_workers           -1.567
year                       NaN
month                   -2.306
day_of_month            -1.576
quarter_Quarter1        -0.643
quarter_Quarter2        -0.627
quarter_Quarter3        -0.470
quarter_Quarter4        -0.507
quarter_Quarter5        -0.203
department_finishing    -0.854
department_sweing       -1.169
day_Monday              -0.451
day_Saturday            -0.419
day_Sunday              -0.460
day_Thursday            -0.431
day_Tuesday             -0.449
day_Wednesday           -0.471
dtype: float64

Train standard deviation after scaling:
team                      0.288
targeted_productivity    10.233
smv                       0.091
wip             

In [142]:
X_reg_train_np = X_reg_train_manual.to_numpy()
X_reg_test_np = X_reg_test_manual.to_numpy()

y_reg_train_np = y_reg_train.to_numpy()
y_reg_test_np = y_reg_test.to_numpy()

print("X train:", X_reg_train_np.shape)
print("X test :", X_reg_test_np.shape)
print("y train:", y_reg_train_np.shape)
print("y test :", y_reg_test_np.shape)

X train: (957, 26)
X test : (240, 26)
y train: (957,)
y test : (240,)


In [143]:
X_reg_train_np = np.c_[
    np.ones(X_reg_train_np.shape[0]),
    X_reg_train_np
]

X_reg_test_np = np.c_[
    np.ones(X_reg_test_np.shape[0]),
    X_reg_test_np
]

print("X train:", X_reg_train_np.shape)
print("X test :", X_reg_test_np.shape)

X train: (957, 27)
X test : (240, 27)


In [146]:
print("NaN values:", np.isnan(X_reg_train_np).sum())
print("Infinite values:", np.isinf(X_reg_train_np).sum())
print("NaN in target:", np.isnan(y_reg_train_np).sum())
print("Infinite in target:", np.isinf(y_reg_train_np).sum())

NaN values: 957
Infinite values: 0
NaN in target: 0
Infinite in target: 0


In [147]:
print("Shape:", X_reg_train_np.shape)
print("Minimum:", np.min(X_reg_train_np))
print("Maximum:", np.max(X_reg_train_np))

Shape: (957, 27)
Minimum: nan
Maximum: nan


### Removing NAN values 

In [149]:
# Start again from the original training and testing data

X_reg_train_manual = X_reg_train.copy()
X_reg_test_manual = X_reg_test.copy()

# 1. Fill missing numerical values using training medians
numeric_medians = X_reg_train_manual[numeric_cols].median()

X_reg_train_manual[numeric_cols] = (
    X_reg_train_manual[numeric_cols]
    .fillna(numeric_medians)
)

X_reg_test_manual[numeric_cols] = (
    X_reg_test_manual[numeric_cols]
    .fillna(numeric_medians)
)

# 2. Remove extra spaces from categorical values
for col in categorical_cols:
    X_reg_train_manual[col] = X_reg_train_manual[col].str.strip()
    X_reg_test_manual[col] = X_reg_test_manual[col].str.strip()

# 3. One-hot encode categorical columns
X_reg_train_manual = pd.get_dummies(
    X_reg_train_manual,
    columns=categorical_cols,
    dtype=float
)

X_reg_test_manual = pd.get_dummies(
    X_reg_test_manual,
    columns=categorical_cols,
    dtype=float
)

# 4. Make sure train and test have exactly the same columns
X_reg_test_manual = X_reg_test_manual.reindex(
    columns=X_reg_train_manual.columns,
    fill_value=0
)

# 5. Calculate mean and standard deviation ONLY from training data
train_mean = X_reg_train_manual.mean()
train_std = X_reg_train_manual.std()

# Avoid division by zero for constant columns
train_std[train_std == 0] = 1

# 6. Scale using training statistics
X_reg_train_manual = (
    X_reg_train_manual - train_mean
) / train_std

X_reg_test_manual = (
    X_reg_test_manual - train_mean
) / train_std

In [150]:
print("NaN in train:", X_reg_train_manual.isna().sum().sum())
print("NaN in test :", X_reg_test_manual.isna().sum().sum())

print("Train shape:", X_reg_train_manual.shape)
print("Test shape :", X_reg_test_manual.shape)

NaN in train: 0
NaN in test : 0
Train shape: (957, 26)
Test shape : (240, 26)


### Convert pandas to numpy

In [151]:
X_reg_train_np = X_reg_train_manual.to_numpy()
X_reg_test_np = X_reg_test_manual.to_numpy()

y_reg_train_np = y_reg_train.to_numpy()
y_reg_test_np = y_reg_test.to_numpy()

In [152]:
print("NaN values:", np.isnan(X_reg_train_np).sum())
print("Infinite values:", np.isinf(X_reg_train_np).sum())
print("Shape:", X_reg_train_np.shape)

NaN values: 0
Infinite values: 0
Shape: (957, 26)


### Adding Intercept 

In [153]:
X_reg_train_np = np.c_[
    np.ones(X_reg_train_np.shape[0]),
    X_reg_train_np
]

X_reg_test_np = np.c_[
    np.ones(X_reg_test_np.shape[0]),
    X_reg_test_np
]

print("Train shape:", X_reg_train_np.shape)
print("Test shape :", X_reg_test_np.shape)

Train shape: (957, 27)
Test shape : (240, 27)


In [154]:
start_time = time.perf_counter()

theta = np.linalg.pinv(
    X_reg_train_np.T @ X_reg_train_np
) @ X_reg_train_np.T @ y_reg_train_np

reg_manual_train_time = time.perf_counter() - start_time

print("Training time:", reg_manual_train_time)
print("Theta shape:", theta.shape)

Training time: 0.0017058000667020679
Theta shape: (27,)


### Calculation Prediction Time

In [155]:
start_time = time.perf_counter()

y_reg_pred_manual = X_reg_test_np @ theta

reg_manual_prediction_time = time.perf_counter() - start_time

print("Prediction time:", reg_manual_prediction_time)
print("First 10 predictions:")
print(y_reg_pred_manual[:10])

Prediction time: 6.240000948309898e-05
First 10 predictions:
[0.80867886 0.7738111  0.77116082 0.74246164 0.87257341 0.72709124
 0.68250385 0.72885258 0.7835859  0.72527431]


### Calculating Prediction Error

In [156]:
reg_manual_mae = np.mean(
    np.abs(y_reg_test_np - y_reg_pred_manual)
)

print("Manual MAE:", reg_manual_mae)

Manual MAE: 0.10545491172373739


In [157]:
reg_manual_rmse = np.sqrt(
    np.mean(
        (y_reg_test_np - y_reg_pred_manual) ** 2
    )
)

print("Manual RMSE:", reg_manual_rmse)

Manual RMSE: 0.14422873696458005


In [158]:
ss_res = np.sum(
    (y_reg_test_np - y_reg_pred_manual) ** 2
)

ss_total = np.sum(
    (y_reg_test_np - np.mean(y_reg_test_np)) ** 2
)

reg_manual_r2 = 1 - (ss_res / ss_total)

print("Manual R2:", reg_manual_r2)

Manual R2: 0.32500505668903257


In [159]:
reg_manual_results = pd.DataFrame({
    "Model": ["Manual Linear Regression"],
    "MAE": [reg_manual_mae],
    "RMSE": [reg_manual_rmse],
    "R2": [reg_manual_r2],
    "Training Time": [reg_manual_train_time],
    "Prediction Time": [reg_manual_prediction_time]
})

reg_manual_results

,Model,MAE,RMSE,R2,Training Time,Prediction Time
0,Manual Linear Regression,0.105455,0.144229,0.325005,0.001706,0.000062


### FOR LOGISTIC REGRESSION 

### Make copies

In [160]:
X_cls_train_manual = X_cls_train.copy()
X_cls_test_manual = X_cls_test.copy()

In [161]:
numeric_medians_cls = X_cls_train_manual[numeric_cols].median()

X_cls_train_manual[numeric_cols] = (
    X_cls_train_manual[numeric_cols]
    .fillna(numeric_medians_cls)
)

X_cls_test_manual[numeric_cols] = (
    X_cls_test_manual[numeric_cols]
    .fillna(numeric_medians_cls)
)

In [162]:
print("Missing values in train:",
    X_cls_train_manual.isna().sum().sum())

print("Missing values in test:",
    X_cls_test_manual.isna().sum().sum())

Missing values in train: 0
Missing values in test: 0


In [163]:
for col in categorical_cols:
    X_cls_train_manual[col] = X_cls_train_manual[col].str.strip()
    X_cls_test_manual[col] = X_cls_test_manual[col].str.strip()

In [164]:
X_cls_train_manual = pd.get_dummies(
    X_cls_train_manual,
    columns=categorical_cols,
    dtype=float
)

X_cls_test_manual = pd.get_dummies(
    X_cls_test_manual,
    columns=categorical_cols,
    dtype=float
)

In [165]:
X_cls_test_manual = X_cls_test_manual.reindex(
    columns=X_cls_train_manual.columns,
    fill_value=0
)

print("Train shape:", X_cls_train_manual.shape)
print("Test shape :", X_cls_test_manual.shape)

print(
    "Same columns:",
    list(X_cls_train_manual.columns) ==
    list(X_cls_test_manual.columns)
)

Train shape: (957, 26)
Test shape : (240, 26)
Same columns: True


### Calculate training mean and standard deviation

In [166]:
train_mean_cls = X_cls_train_manual.mean()
train_std_cls = X_cls_train_manual.std()

# Avoid division by zero for constant columns
train_std_cls[train_std_cls == 0] = 1

In [167]:
X_cls_train_manual = (
    X_cls_train_manual - train_mean_cls
) / train_std_cls

X_cls_test_manual = (
    X_cls_test_manual - train_mean_cls
) / train_std_cls

### Checking for NAN

In [168]:
print("NaN in train:",
        X_cls_train_manual.isna().sum().sum())

print("NaN in test:",
        X_cls_test_manual.isna().sum().sum())

print("Train shape:", X_cls_train_manual.shape)
print("Test shape :", X_cls_test_manual.shape)

NaN in train: 0
NaN in test: 0
Train shape: (957, 26)
Test shape : (240, 26)


### Converting into numpy 

In [169]:
X_cls_train_np = X_cls_train_manual.to_numpy()
X_cls_test_np = X_cls_test_manual.to_numpy()

y_cls_train_np = y_cls_train.to_numpy()
y_cls_test_np = y_cls_test.to_numpy()

In [170]:
print("X train:", X_cls_train_np.shape)
print("X test :", X_cls_test_np.shape)
print("y train:", y_cls_train_np.shape)
print("y test :", y_cls_test_np.shape)

X train: (957, 26)
X test : (240, 26)
y train: (957,)
y test : (240,)


In [171]:
X_cls_train_np = np.c_[
    np.ones(X_cls_train_np.shape[0]),
    X_cls_train_np
]

X_cls_test_np = np.c_[
    np.ones(X_cls_test_np.shape[0]),
    X_cls_test_np
]

In [172]:
print("X train:", X_cls_train_np.shape)
print("X test :", X_cls_test_np.shape)

X train: (957, 27)
X test : (240, 27)


### Creating the Sigmoid Function 

In [174]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

theta_cls = np.zeros(X_cls_train_np.shape[1])

learning_rate = 0.01
iterations = 5000

### Gradient descent

In [175]:
start_time = time.perf_counter()

n = X_cls_train_np.shape[0]

for i in range(iterations):

    # Linear score
    z = X_cls_train_np @ theta_cls

    # Convert score to probability
    prob = sigmoid(z)

    # Calculate gradient
    gradient = (
        X_cls_train_np.T @ (prob - y_cls_train_np)
    ) / n

    # Update coefficients
    theta_cls = theta_cls - learning_rate * gradient

cls_manual_train_time = time.perf_counter() - start_time

print("Training time:", cls_manual_train_time)
print("Theta shape:", theta_cls.shape)

Training time: 0.08984780008904636
Theta shape: (27,)


### Predict probabilities

In [176]:
start_time = time.perf_counter()

z_test = X_cls_test_np @ theta_cls

y_cls_prob_manual = sigmoid(z_test)

cls_manual_prediction_time = time.perf_counter() - start_time

print("Prediction time:", cls_manual_prediction_time)
print("First 10 probabilities:")
print(y_cls_prob_manual[:10])

Prediction time: 0.00011030002497136593
First 10 probabilities:
[0.67424319 0.62705308 0.63290959 0.373396   0.94415041 0.86644038
 0.89026286 0.79608538 0.89288775 0.57115273]


### Convert Probabilities to Predictions

In [177]:
y_cls_pred_manual = (
    y_cls_prob_manual >= 0.5
).astype(int)

print("First 10 predictions:")
print(y_cls_pred_manual[:10])

First 10 predictions:
[1 1 1 0 1 1 1 1 1 1]


### Calculating Manual Accuracy 

In [178]:
correct = np.sum(
    y_cls_test_np == y_cls_pred_manual
)

total = len(y_cls_test_np)

cls_manual_accuracy = correct / total

print("Manual Accuracy:", cls_manual_accuracy)

Manual Accuracy: 0.75


### Calculate the confusion counts

In [179]:
tp = np.sum(
    (y_cls_test_np == 1) &
    (y_cls_pred_manual == 1)
)

tn = np.sum(
    (y_cls_test_np == 0) &
    (y_cls_pred_manual == 0)
)

fp = np.sum(
    (y_cls_test_np == 0) &
    (y_cls_pred_manual == 1)
)

fn = np.sum(
    (y_cls_test_np == 1) &
    (y_cls_pred_manual == 0)
)

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

TP: 161
TN: 19
FP: 50
FN: 10


In [ ]:
cls_manual_precision = tp / (tp + fp)
print("Manual Precision:", cls_manual_precision)

Manual Precision: 0.7630331753554502


In [181]:
cls_manual_recall = tp / (tp + fn)
print("Manual Recall:", cls_manual_recall)

Manual Recall: 0.9415204678362573


In [182]:
cls_manual_f1 = (
    2 * cls_manual_precision * cls_manual_recall
) / (
    cls_manual_precision + cls_manual_recall
)

print("Manual F1:", cls_manual_f1)

Manual F1: 0.8429319371727748


### Saving Final Classification Results

In [183]:
cls_manual_results = pd.DataFrame({
    "Model": ["Manual Logistic Regression"],
    "Accuracy": [cls_manual_accuracy],
    "Precision": [cls_manual_precision],
    "Recall": [cls_manual_recall],
    "F1": [cls_manual_f1],
    "Training Time": [cls_manual_train_time],
    "Prediction Time": [cls_manual_prediction_time]
})

cls_manual_results

,Model,Accuracy,Precision,Recall,F1,Training Time,Prediction Time
0,Manual Logistic Regression,0.75,0.763033,0.94152,0.842932,0.089848,0.00011


In [184]:
reg_comparison = pd.concat(
    [reg_results, reg_manual_results],
    ignore_index=True
)

reg_comparison

,Model,MAE,RMSE,R2,Training Time,Prediction Time
0,Scikit-learn Linear Regression,0.105137,0.144063,0.326551,0.002947,0.000465
1,Manual Linear Regression,0.105455,0.144229,0.325005,0.001706,0.000062


In [185]:
cls_comparison = pd.concat(
    [cls_results, cls_manual_results],
    ignore_index=True
)

cls_comparison

,Model,Accuracy,Precision,Recall,F1,Training Time,Prediction Time
0,Scikit-learn Logistic Regression,0.7625,0.776699,0.935673,0.848806,0.010902,0.000309
1,Manual Logistic Regression,0.7500,0.763033,0.941520,0.842932,0.089848,0.000110


## Part C - Comparison and Optimization

### Optimizing with learning rate 0.01

In [196]:
theta_cls_improved = np.zeros(X_cls_train_np.shape[1])

learning_rate_improved = 0.01
iterations_improved = 10000

start_time = time.perf_counter()

n = X_cls_train_np.shape[0]

for i in range(iterations_improved):

    z = X_cls_train_np @ theta_cls_improved

    prob = sigmoid(z)

    gradient = (
        X_cls_train_np.T @ (prob - y_cls_train_np)
    ) / n

    theta_cls_improved = (
        theta_cls_improved -
        learning_rate_improved * gradient
    )

cls_improved_train_time = time.perf_counter() - start_time

print("Improved training time:", cls_improved_train_time)

Improved training time: 0.1730326000833884


In [193]:
start_time = time.perf_counter()

z_test_improved = X_cls_test_np @ theta_cls_improved

y_cls_prob_improved = sigmoid(z_test_improved)

y_cls_pred_improved = (
    y_cls_prob_improved >= 0.5
).astype(int)

cls_improved_prediction_time = time.perf_counter() - start_time

print("Improved prediction time:",
        cls_improved_prediction_time)

Improved prediction time: 0.00012420001439750195


In [194]:
correct_improved = np.sum(
    y_cls_test_np == y_cls_pred_improved
)

cls_improved_accuracy = (
    correct_improved / len(y_cls_test_np)
)

print("Improved Accuracy:", cls_improved_accuracy)

Improved Accuracy: 0.7541666666666667


### Optimizing with learning rate 0.05 

In [197]:
theta_cls_improved = np.zeros(X_cls_train_np.shape[1])

learning_rate_improved = 0.05
iterations_improved = 10000

start_time = time.perf_counter()

n = X_cls_train_np.shape[0]

for i in range(iterations_improved):

    z = X_cls_train_np @ theta_cls_improved

    prob = sigmoid(z)

    gradient = (
        X_cls_train_np.T @ (prob - y_cls_train_np)
    ) / n

    theta_cls_improved = (
        theta_cls_improved -
        learning_rate_improved * gradient
    )

cls_improved_train_time = time.perf_counter() - start_time

print("Improved training time:", cls_improved_train_time)

Improved training time: 0.17505789990536869


In [198]:
start_time = time.perf_counter()

z_test_improved = X_cls_test_np @ theta_cls_improved

y_cls_prob_improved = sigmoid(z_test_improved)

y_cls_pred_improved = (
    y_cls_prob_improved >= 0.5
).astype(int)

cls_improved_prediction_time = time.perf_counter() - start_time

print("Improved prediction time:",
        cls_improved_prediction_time)

Improved prediction time: 0.0003961999900639057


In [200]:
correct_improved = np.sum(
    y_cls_test_np == y_cls_pred_improved
)

cls_improved_accuracy = (
    correct_improved / len(y_cls_test_np)
)

print("Improved Accuracy:", cls_improved_accuracy)

Improved Accuracy: 0.7666666666666667


### Confusion matrix counts

In [201]:
tp = np.sum(
    (y_cls_test_np == 1) &
    (y_cls_pred_improved == 1)
)

tn = np.sum(
    (y_cls_test_np == 0) &
    (y_cls_pred_improved == 0)
)

fp = np.sum(
    (y_cls_test_np == 0) &
    (y_cls_pred_improved == 1)
)

fn = np.sum(
    (y_cls_test_np == 1) &
    (y_cls_pred_improved == 0)
)

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

TP: 160
TN: 24
FP: 45
FN: 11


In [202]:
improved_precision = tp / (tp + fp)
print("Improved Precision:", improved_precision)

Improved Precision: 0.7804878048780488


In [203]:
improved_recall = tp / (tp + fn)
print("Improved Recall:", improved_recall)

Improved Recall: 0.935672514619883


In [204]:
improved_f1 = (
    2 * improved_precision * improved_recall
) / (
    improved_precision + improved_recall
)
print("Improved F1:", improved_f1)


Improved F1: 0.8510638297872339


In [205]:
cls_improved_results = pd.DataFrame({
    "Model": ["Optimized Manual Logistic Regression"],
    "Accuracy": [cls_improved_accuracy],
    "Precision": [improved_precision],
    "Recall": [improved_recall],
    "F1": [improved_f1],
    "Training Time": [cls_improved_train_time],
    "Prediction Time": [cls_improved_prediction_time]
})

cls_improved_results

,Model,Accuracy,Precision,Recall,F1,Training Time,Prediction Time
0,Optimized Manual Logistic Regression,0.766667,0.780488,0.935673,0.851064,0.181816,0.000396


In [206]:
final_cls_comparison = pd.concat(
    [cls_results, cls_manual_results, cls_improved_results],
    ignore_index=True
)

final_cls_comparison

,Model,Accuracy,Precision,Recall,F1,Training Time,Prediction Time
0,Scikit-learn Logistic Regression,0.762500,0.776699,0.935673,0.848806,0.010902,0.000309
1,Manual Logistic Regression,0.750000,0.763033,0.941520,0.842932,0.089848,0.000110
2,Optimized Manual Logistic Regression,0.766667,0.780488,0.935673,0.851064,0.181816,0.000396


In [207]:
final_reg_comparison = pd.concat(
    [reg_results, reg_manual_results],
    ignore_index=True
)
final_reg_comparison

,Model,MAE,RMSE,R2,Training Time,Prediction Time
0,Scikit-learn Linear Regression,0.105137,0.144063,0.326551,0.002947,0.000465
1,Manual Linear Regression,0.105455,0.144229,0.325005,0.001706,0.000062


### Observations

For regression, the manual Linear Regression implementation produced results very close to the Scikit-learn implementation. The manual model obtained an MAE of 0.105455, RMSE of 0.144229, and R² of 0.325005, compared with 0.105137, 0.144063, and 0.326551 for Scikit-learn. This shows that the manually implemented closed-form solution can reproduce the predictive performance of the library implementation closely.

For classification, the baseline manual Logistic Regression achieved an accuracy of 0.7500 and an F1 score of 0.842932. Increasing the number of iterations from 5,000 to 10,000 produced only a small improvement. After increasing the learning rate from 0.01 to 0.05 while keeping 10,000 iterations, the optimized manual model achieved an accuracy of 0.766667 and an F1 score of 0.851064. This was slightly higher than the Scikit-learn Logistic Regression accuracy of 0.7625 and F1 score of 0.848806.

The manual implementations were able to achieve predictive performance close to or slightly better than Scikit-learn in these experiments. However, the manual Logistic Regression required more training time than the Scikit-learn implementation because the optimization loop was implemented explicitly in Python. The prediction times remained very small for both approaches.

### Conclusion

The experiment demonstrates that machine learning algorithms such as Linear Regression and Logistic Regression can be implemented from scratch using NumPy and Pandas while achieving performance comparable to Scikit-learn. Vectorized NumPy operations made the manual implementations efficient, while preprocessing ensured that missing values, categorical features, and feature scales were handled consistently.

The classification experiment also showed the effect of optimization parameters on gradient descent. Increasing the learning rate from 0.01 to 0.05 improved the manual Logistic Regression performance, demonstrating the importance of choosing suitable optimization settings.

Overall, Scikit-learn provides a convenient and optimized implementation with lower training overhead, while the manual implementations provide a better understanding of the underlying mathematical operations and training process.
